zipline/gens/brokers/alpaca_broker.py

```diff
@@ -0,0 +1,261 @@
+
+        order = self._api.submit_order(
+            symbol=symbol,
+            qty=qty,
+            side=side,
+            type=order_type,
+            time_in_force='day',
+            limit_price=limit_price,
+            stop_price=stop_price,
+            client_order_id=zp_order.id,
+        )
+        zp_order = self._order2zp(order)
+        return zp_order
+
+    @property
+    def orders(self):
+        return {
+            o.client_order_id: self._order2zp(o)
+            for o in self._api.list_orders()
+        }
+
+    @property
+    def transactions(self):
+        return {
+            o.client_order_id: self._order2zp(o)
+            for o in self._api.list_orders(status='all')
+        }
+
+    def cancel_order(self, zp_order_id):
+        try:
+            order = self._api.get_order_by_client_order_id(zp_order_id)
+            self._api.cancel_order(order.id)
+        except Exception as e:
+            log.error(e)
+            return
+
+    def get_last_traded_dt(self, asset):
+        quote = self._api.get_quote(asset.symbol)
+        return pd.Timestamp(quote.last_timestamp)
+
+    def get_spot_value(self, assets, field, dt, data_frequency):
+        assert(field in (
+            'open', 'high', 'low', 'close', 'volume', 'price', 'last_traded'))
+        assets_is_scalar = not isinstance(assets, (list, set, tuple))
+        is_daily = 'd' in data_frequency
+        if assets_is_scalar:
+            symbols = [assets.symbol]
+        else:
+            symbols = [asset.symbol for asset in assets]
+        if field == 'price' or field == 'last_traded':
+            quotes = self._api.list_quotes(symbols)
+            if assets_is_scalar:
+                if field == 'price':
+                    return quotes[-1].last
+                else:
+                    return quotes[-1].last_timestamp
+            else:
+                return [
+                    quote.last if field == 'price' else quote.last_timestamp
+                    for quote in quotes
+                ]
+
+        timeframe = '1D' if is_daily else '1Min'
+        bars_list = self._api.list_bars(symbols, timeframe, limit=1)
+        if assets_is_scalar:
+            return bars_list[0].bars[-1]._raw[field]
+        bars_map = {a.symbol: a for a in bars_list}
+        return [
+            bars_map[symbol].bars[-1]._raw[field]
+            for symbol in symbols
+        ]
+
+    def get_realtime_bars(self, assets, data_frequency):
+        # TODO: cache the result. The caller
+        # (DataPortalLive#get_history_window) makes use of only one
+        # column at a time.
+        assets_is_scalar = not isinstance(assets, (list, set, tuple))
+        is_daily = 'd' in data_frequency
+        if assets_is_scalar:
+            symbols = [assets.symbol]
+        else:
+            symbols = [asset.symbol for asset in assets]
+        timeframe = '1D' if is_daily else '1Min'
+
+        bars_list = self._api.list_bars(symbols, timeframe, limit=500)
```

Comment:
Why 500?

Reply:
I wanted to get the actual bars number from upper caller, but seems like this API doesn't get bar counts.  That's rather weird to me and I wonder how ib_broker gets bars. 500 is just a heuristic number.